### PR TITLE
Added support for ZCl reply

### PR DESCRIPTION
**Pull request by AndreasBomholtz**
I needed ZCl reply to test OTA with my devices.

### CODE REVIEW

**Comment by rcloran** [comment]
8fd9d55ef93df72cbd4227f1ba6de47992b5c052 should fix the build issues you were seeing here.

File: bellows/zigbee/zcl/__init__.py [hunk 1]
```diff
@@ -77,13 +77,19 @@ def __init__(cls, name, bases, nmspc):
             for command_id, details in cls.server_commands.items():
                 command_name, schema, is_reply = details
                 cls._server_command_idx[command_name] = command_id
+        if hasattr(cls, 'client_commands'):
```
Comment (rcloran) [inline bellows/zigbee/zcl/__init__.py]: Can you put the client_commands stuff in a separate PR, please?

Comment (rcloran) [reply bellows/zigbee/zcl/__init__.py]: Actually, never mind ... this is OK as is.

Comment (AndreasBomholtz) [reply bellows/zigbee/zcl/__init__.py]: Okay

**Comment by AndreasBomholtz** [comment]
Yes I have added OTA support for my program and that is how I noticed that reply was missing.
You have the "handle_cluster_request" in zcl/cluster/__init__.py, can we use that to add the OTA handling directly to the cluster or ...?

File: bellows/zigbee/application.py [hunk 2]
```diff
@@ -263,8 +264,22 @@ def request(self, nwk, aps_frame, data, timeout=10):
...
 
+    @asyncio.coroutine
     def reply(self, nwk, aps_frame, data):
-        return self._ezsp.sendUnicast(self.direct, nwk, aps_frame, aps_frame.sequence, data)
+        seq = aps_frame.sequence
+        assert seq not in self._pending
+        send_fut = asyncio.Future()
+        self._pending[seq] = (send_fut, None)
+
+        v = yield from self._ezsp.sendUnicast(self.direct, nwk, aps_frame, seq, data)
```
[the remark this answers is not in the document]
Comment (AndreasBomholtz) [reply bellows/zigbee/application.py]: I will just try to play with that and see how it behaviors.

[the remark this answers is not in the document]
Comment (AndreasBomholtz) [reply bellows/zigbee/application.py]: I have just tried this out, and it do not work well :)
The policy UNICAST_REPLIES_POLICY was to be changed to HOST_WILL_SUPPLY_REPLY, which means that bellows has to send a reply for all the incoming unicast frames. The default mode is that the NCP handles the replies.
When I run the OTA process with sendUnicast it works fine.

[the remark this answers is not in the document]
Comment (rcloran) [reply bellows/zigbee/application.py]: Thanks for looking into that -- I guess I can imagine a future world where bellows might handle all replies like that, but the need isn't clear.

**Comment by rcloran** [comment]
Trying to think through this ... Please do point out where my understanding is wrong.

 - We currently use `command()` for client -> server messages
 - We need to support server -> client messages, too
   - The most important thing there is setting the direction bit appropriately
 - Most server -> client messages are replies to client -> server (but that's not always true)

As it currently stands, the new `reply()` API encodes two lower level intentions: "send a server -> client message" and "don't expect a reply". That means that we can't send server -> client messages that expect a reply. The only example I've found of a server -> client message that is not actually a reply is OTA cluster's "notify" command (0), and that happens to not expect a response.

I think my initial intent with adding the `reply` methods through the stack was to support proper replies, where the APS sequence matched the request. Since it's broken, I think it should be removed.

What do you think about removing `application.py`'s `reply` method, and instead adding an `expect_reply` argument (or some other better name)? Then, zcl's `request` could determine the direction automatically, based on whether it is a client or server cluster, and client or server command it's sending, and also take an `expect_reply`.

My main concern with this PR is that the new methods duplicate a bunch of code in the existing `request` methods, and I think my proposal fixes that and keeps the API fairly simple.

**Comment by AndreasBomholtz** [comment]
I think that sounds really good. I will change it.

**Comment by rcloran** [comment]
It's not clear to me why you added the ensure_future() inside the `reply` methods. I'm guessing you were using it somewhere that didn't await/yield from it. As they currently stand, the `reply` methods should be treated as a coroutine, even though they're not labeled as such. I'll experiment a bit with a different formulation that doesn't use ensure_future (which was breaking the tests) but still makes it clear that reply is a coroutine.
